src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs

```diff
@@ -1,14 +1,206 @@
+         _headers = new Cache<string, ColumnDescription>();
+         _rawDataTable = new List<List<string>>();
+      }
+
+      private IEnumerable<string> getColumn(int columnIndex)
+      {
+         return _rawDataTable.Select(column => column[columnIndex]).ToList();
+      }
+
+      public void AddColumn(string columnName, int columnIndex) //it seems to me there is little sense in adding column after column
+         //the list of headers is somehow the definition of the table
+      {
+         if (columnName.IsNullOrEmpty())
+         {
+            columnName = Guid.NewGuid().ToString();
+            _emptyColumns.Add(columnName);
+         }
+         _headers.Add(columnName, new ColumnDescription(columnIndex));
+      }
+
+      public void CalculateColumnDescription(List<ColumnDescription.MeasurementLevel> levels)
+      {
+         _headers.Each(header =>
+         {
+            if (header.Level == ColumnDescription.MeasurementLevel.NotSet)
+               header.Level = levels[header.Index];
+         });
+      }
+
+      public void AddRow( IEnumerable<string> row)
+      {
+         var rowList = row.ToList();
+
+         if (_headers.Count > rowList.Count)
+         {
+            for ( var i = rowList.Count; i < _headers.Count; i++  )
+               rowList.Add("");
+         }
+         if (rowList.Count > _headers.Count)
+         {
+            rowList = rowList.GetRange(0, _headers.Count);
+         }
+
+         _rawDataTable.Add(rowList);
+
+         foreach (var header in _headers)
+         {
+            if (!header.ExistingValues.Contains(rowList.ElementAt(header.Index)))
+               header.AddExistingValues(rowList.ElementAt(header.Index));
+         }
+      }
+
+      public IEnumerable<string> GetColumn(string columnName)
+      {
+         return getColumn(_headers[columnName].Index);
+      }
+
+      public IEnumerable<UnformattedRow> GetRows(Func<IEnumerable<string>, bool> filter)
+      {
+         return _rawDataTable.Select((data, index) => new UnformattedRow(index, data)).Where(row => filter(row.Data));
+      }
+
+      public DataTable AsDataTable()
+      {
+         var resultTable = new DataTable();
+         var indexList = _headers.Select( h => h.Index);
+
+         // Add columns.
+         foreach (var header in _headers.Keys)
+         {
+            resultTable.AddColumn(header, typeof(string));
+         }
+
+         Func<IEnumerable<string>, IEnumerable<string>> maskFunction;
+
+         if (_rawDataTable.Count > 0 && (_headers.Count != _rawDataTable.First().Count))
+         {
+            maskFunction = (inputList) => inputList.Where((v, i) => indexList.Contains(i));
+         }
+         else
+         {
+            maskFunction = (inputList) => inputList;
+         }
+
+         foreach (var itemList in _rawDataTable)
+         {
+            // ReSharper disable once CoVariantArrayConversion
+            resultTable.Rows.Add(maskFunction(itemList).ToArray());
```

Comment:
are you sure about this. I have seen cases where this creates crash at run time

Reply:
used .ToArray<object>()explicitely, no warning now and it works